### PR TITLE
[FLINK-6552] Allow differing types for side outputs

### PR DESCRIPTION
**Pull request by aljoscha**
R: @zentol

### CODE REVIEW

**Comment by aljoscha** [comment]
You're right. 😓 I'm fixing and then merging.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java [hunk 1]
```diff
@@ -318,9 +318,10 @@ public void addVirtualSideOutputNode(Integer originalId, Integer virtualId, Outp
 				continue;
 			}
 
-			if (!tag.f1.getTypeInfo().equals(outputTag.getTypeInfo())) {
-				throw new IllegalArgumentException("Trying to add a side input for the same id " +
-						"with a different type. This is not allowed.");
+			if (tag.f1.getId().equals(outputTag.getId()) &&
+					!tag.f1.getTypeInfo().equals(outputTag.getTypeInfo())) {
+				throw new IllegalArgumentException("Trying to add a side output for the same" +
+						"side-output id with a different type. This is not allowed.");
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java]: Could also include the tag name for good measure.

**Comment by aljoscha** [comment]
Merged

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java [hunk 2]
```diff
@@ -318,9 +318,10 @@ public void addVirtualSideOutputNode(Integer originalId, Integer virtualId, Outp
 				continue;
 			}
 
-			if (!tag.f1.getTypeInfo().equals(outputTag.getTypeInfo())) {
-				throw new IllegalArgumentException("Trying to add a side input for the same id " +
-						"with a different type. This is not allowed.");
+			if (tag.f1.getId().equals(outputTag.getId()) &&
+					!tag.f1.getTypeInfo().equals(outputTag.getTypeInfo())) {
+				throw new IllegalArgumentException("Trying to add a side output for the same" +
```
Comment (tedyu) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java]: minor: missing space after "same"

**Comment by aljoscha** [comment]
Thanks @tedyu and @zentol! I pushed another commit with these minor fixes.

**Pull request closed without merging**